File: modules/device-health/android/src/main/java/expo/modules/devicehealth/DeviceHealthModule.kt
package expo.modules.devicehealth

import android.net.TrafficStats
import android.os.Process
import android.os.SystemClock
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.HttpURLConnection
import java.net.URL

class DeviceHealthModule : Module() {

  /**
   * Previous snapshot for delta calculation.
   * Uses Process.getElapsedCpuTime() (app-specific, no /proc/stat permission needed).
   */
  private var prevCpuTimeMs: Long = 0
  private var prevWallTimeMs: Long = 0
  private var hasPreviousSnapshot = false

  override fun definition() = ModuleDefinition {
    Name("DeviceHealth")

    AsyncFunction("getCPUUsage") {
      val cpuTimeMs = Process.getElapsedCpuTime()   // ms of CPU consumed by this process
      val wallTimeMs = SystemClock.elapsedRealtime() // ms since boot (monotonic)
      val coreCount = Runtime.getRuntime().availableProcessors()

      var usagePercent = -1.0

      if (hasPreviousSnapshot) {
        val cpuDelta = cpuTimeMs - prevCpuTimeMs
        val wallDelta = wallTimeMs - prevWallTimeMs

        if (wallDelta > 0 && coreCount > 0) {
          // Normalize: cpuDelta is across all cores, so divide by coreCount for 0-100% range
          usagePercent = (cpuDelta.toDouble() / (wallDelta * coreCount)) * 100.0
          usagePercent = Math.round(usagePercent * 10.0) / 10.0
          // Clamp to 0-100 (can exceed 100 on burst scheduling)
          usagePercent = usagePercent.coerceIn(0.0, 100.0)
        }
      }

      prevCpuTimeMs = cpuTimeMs
      prevWallTimeMs = wallTimeMs
      hasPreviousSnapshot = true

      mapOf(
        "usagePercent" to usagePercent,
        "coreCount" to coreCount
      )
    }

    // ── Network Traffic Stats ──────────────────────────────────────

    /**
     * Read cumulative network byte counters for this app's UID.
     * Uses TrafficStats which reads from /proc/net/xt_qtaguid/stats.
     * Returns { bytesReceived, bytesSent } — cumulative since device boot.
     * JS layer calculates delta between polls to get throughput.
     */
    Function("getNetworkStats") {
      val uid = Process.myUid()
      val rxBytes = TrafficStats.getUidRxBytes(uid)
      val txBytes = TrafficStats.getUidTxBytes(uid)

      mapOf(
        "bytesReceived" to if (rxBytes != TrafficStats.UNSUPPORTED.toLong()) rxBytes else -1L,
        "bytesSent" to if (txBytes != TrafficStats.UNSUPPORTED.toLong()) txBytes else -1L
      )
    }

    // ── Speed Test ──────────────────────────────────────────────────

    /**
     * Download a file from a URL and measure throughput.
     * Downloads the entire response for accurate measurement.
     * Returns { downloadSpeedMbps, bytesDownloaded, durationMs }
     */
    AsyncFunction("runSpeedTest") { testUrl: String ->
      val connectTimeoutMs = 15_000
      val readTimeoutMs = 15_000

      var connection: HttpURLConnection? = null

      try {
        val url = URL(testUrl)
        connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        connection.connectTimeout = connectTimeoutMs
        connection.readTimeout = readTimeoutMs
        connection.useCaches = false

        val startTime = SystemClock.elapsedRealtime()
        connection.connect()

        val inputStream = connection.inputStream
        val buffer = ByteArray(8192)
        var totalBytesRead = 0

        while (true) {
          val bytesRead = inputStream.read(buffer)
          if (bytesRead == -1) break
          totalBytesRead += bytesRead
        }

        inputStream.close()
        val endTime = SystemClock.elapsedRealtime()
        val durationMs = endTime - startTime

        var speedMbps = -1.0
        if (totalBytesRead > 0 && durationMs > 0) {
          speedMbps = (totalBytesRead.toDouble() * 8.0) / (durationMs * 1000.0)
          speedMbps = Math.round(speedMbps * 100.0) / 100.0
        }

        mapOf(
          "downloadSpeedMbps" to speedMbps,
          "bytesDownloaded" to totalBytesRead,
          "durationMs" to durationMs.toDouble()
        )
      } catch (e: Exception) {
        android.util.Log.w("DeviceHealth", "Speed test failed: ${e.message}")
        mapOf(
          "downloadSpeedMbps" to -1.0,
          "bytesDownloaded" to 0,
          "durationMs" to -1.0
        )
      } finally {
        try { connection?.disconnect() } catch (_: Exception) {}
      }
    }

    /** Reset stored snapshot (call when monitoring session ends) */
    Function("resetCPUSnapshot") {
      hasPreviousSnapshot = false
      prevCpuTimeMs = 0
      prevWallTimeMs = 0
    }
  }
}

File: modules/device-health/src/DeviceHealthModule.ts
import { requireNativeModule } from 'expo-modules-core';

export interface CPUUsage {
  /** CPU usage percentage (0-100). Returns -1 on first call (no previous snapshot to compare). */
  usagePercent: number;
  /** Number of active CPU cores */
  coreCount: number;
}

export interface NetworkStats {
  /** Cumulative bytes received by the app since device boot. -1 if unsupported. */
  bytesReceived: number;
  /** Cumulative bytes sent by the app since device boot. -1 if unsupported. */
  bytesSent: number;
}

const DeviceHealth = requireNativeModule('DeviceHealth');

/**
 * Get current system-wide CPU usage as a percentage.
 *
 * Uses delta between consecutive calls — first call returns -1 (priming read).
 * Recommended polling interval: 3-5 seconds.
 *
 * iOS: Mach kernel host_statistics (HOST_CPU_LOAD_INFO)
 * Android: /proc/stat tick counters
 */
export async function getCPUUsage(): Promise<CPUUsage> {
  return DeviceHealth.getCPUUsage();
}

/**
 * Reset the stored CPU tick snapshot.
 * Call when a monitoring session ends so the next session starts fresh.
 */
export function resetCPUSnapshot(): void {
  DeviceHealth.resetCPUSnapshot();
}

/**
 * Read cumulative network byte counters from OS.
 * Returns bytes received/sent since device boot — JS calculates delta for throughput.
 *
 * iOS: getifaddrs() — sums all network interface byte counters
 * Android: TrafficStats.getUidRxBytes/getUidTxBytes — app-specific counters
 *
 * This is a synchronous native call (no network I/O), safe to poll every 5s.
 */
export function getNetworkStats(): NetworkStats {
  return DeviceHealth.getNetworkStats();
}

export interface SpeedTestResult {
  /** Measured download speed in Mbps. -1 if failed. */
  downloadSpeedMbps: number;
  /** Bytes downloaded during the test. */
  bytesDownloaded: number;
  /** Total test duration in ms. -1 if failed. */
  durationMs: number;
}

/**
 * Run a one-time speed test by downloading a file from a URL.
 * Downloads the entire file (~3.5MB) for accurate throughput measurement.
 *
 * iOS: URLSession async download (no caching)
 * Android: HttpURLConnection full download
 *
 * @param testUrl - Fixed URL to a known-size file on the app's CDN
 */
export async function runSpeedTest(testUrl: string): Promise<SpeedTestResult> {
  return DeviceHealth.runSpeedTest(testUrl);
}
